### ISSUE TITLE
Feature request: use extension without keywords

### ISSUE THREAD

**Issue by nicoulaj** [Ulauncher/Ulauncher#306]
Some extensions could be used without a trigger keyword, for instance:

 * system control: `reboot`, etc
 * media player control: `play`, `pause`, etc

This should be configurable by the user on a per extension basis, eg by setting the keyword empty.

As an example, rofi supports this (combi mode).

**Comment by friday** [comment]
Hey again,

I agree this would be at least a "nice-to-have". It could be used to allow unit and/or currency conversion unprefixed (similar to how Ulauncher implements the calculator now). In addition I have seen someone request unprefixed file search.

Your examples are not impossible to implement though. You could use multiple keywords instead. I've done something similar here: https://github.com/friday/ulauncher-gnome-settings

**Comment by friday** [comment]
Also, perhaps this issue could be clarified into one of either:

1. Support extensions that work like Apps (no ExtensionResultItem, arguments or [space](https://github.com/Ulauncher/Ulauncher/issues/276))
2. Support extensions without keywords (like the built in calculator)

I think this needs to be distinguished since the behavior and implementation details are very different. I also think 1 is more important and realistic, and also is what you want here?

**Comment by nicoulaj** [comment]
I agree, also use case 1 can be subdivided into:

1.1. Extensions with a static set of keywords: eg player control, system control -> your workarounds apply (although it is not very comfortable having to fork every plugin to do this, so just creating desktop entries seems easier)
1.2. Extensions with a finite, but dynamic set of keywords: eg browser bookmarks, indexed IDE projects [like this one](https://github.com/albertlauncher/python/blob/master/jetbrains-projects.py)

Use case 2 seems harder to implement to me without impacting performance, but I don't know how ulauncher communicates with extensions (threading ? asynchronism ?)

**Comment by friday** [comment]
Extensions use websockets to communicate with Ulauncher, and the extensions are kept running in the background. Performance could still be an issue, but that's not the problem I was thinking of. Ulauncher has a specific flow. You either see the apps and keywords matching your input, or a result list from exactly one extension. Keywordless implies seeing apps in addition to multiple extensions results at the same time.

Extension preferences enables users to rename keywords. Having all of them in the manifest is needed for this. So I don't think adding keywords dynamically is a good idea. Adding a static set of keywords in the extension manifest and disabling dynamically if they can't be supported is better imo.

This could be used in a hashing extension if the corresponding algorithm for a keyword isn't supported by the system for example (I wanted to do this so this is not a hypothetical example).

To match a list of “things“ (like browser bookmarks, desktop apps, commands, files etc) I think 2. or using a wrapper keyword is the way to go. Keywords shouldn't be used for this. My extension is more of a hack to fill a void, than a good and valid use case to support.

**Comment by gornostal** [comment]
@nicoulaj thanks for the feature request. 

I'm planning to review priorities of all of them once we migrate to Python 3.

**Comment by friday** [comment]
For simple commands such as reboot, play and pause, this could be used in the meanwhile: https://github.com/Ulauncher/Ulauncher/issues/191#issuecomment-457000841

**Comment by gornostal** [comment]
I've gave this another thought, I would like to keep keywords mandatory for extensions.
If extension doesn't have a keyword, it means it's a simple executable.
Once #191 is implemented the problem can be easily solved with shortcuts.

**Comment by friday** [comment]
> If extension doesn't have a keyword, it means it's a simple executable.

I don't agree with the motivation. For example, the built in calculator doesn't have an argument, but it's not a simple executable.

Fair enough on the decision though. Too much flexibility is hard to manage, and this would definitely be a complex one to implement.

**Comment by simonwiles** [comment]
I'm trying to migrate from Kupfer, and this is proving a bit frustrating.  In particular for file-search, and I'm used to using extension that add to the search catalog.  In particular, I have begun developing a Ulauncher plugin that allows searching notes stored in the Zim Wiki application, but its really awkward to have to type a keyword.  Clearly Ulauncher and Kupfer have different underlying operating models, but if Ulauncher could support the ability to add items to its internal index, it would allow different kinds of behaviour that cannot currently be supported (and #191 doesn't help this).

**Comment by arnoo** [comment]
I'd really wish you'd reconsider allowing extensions to work without keywords. What makes executables so special that they should be the only thing accessible without keywords ?

For example, I would like to access my bookmarks directly : I don't want to have to think about whether some webapp is an executable or a website every time I want to use it. I have resorted to adding Desktop files for the websites I visit most often, but I still often find myself typing the name of websites in ulauncher without prefix and then adding the prefix afterwards because ulauncher is not finding anything. I think this defeats the purpose of finding things as quickly as possible.

**Comment by KuenzelIT** [comment]
> For example, I would like to access my bookmarks directly.

I totally agree with this. The best thing about the Albert launcher was exactly this behaviour, so intuitive and easy. It didn't matter if I wanted to open a website or a ssh connection. Isn't there any way to this?

**Comment by friday** [comment]
I'm reopening this. I think there's a way to support keywordless shortcuts, and would like to keep an open issue for reference. It won't be easy to support or added anytime soon though.

My thought is that extensions should be able to specify a pattern for valid input. Maybe with regex, or maybe with something more performant than regex. If an extension does that, then users can be given an option in the preferences to make the keyword optional, and drag and drop the extensions to set higher or lower prio.

So if the user installs a currency conversion extension, and makes the `conv` keyword optional, then enter "5 USD to GBP", this triggers the same view as if they entered "`conv` 5 USD to GBP" before.

**Comment by KuenzelIT** [comment]
Sounds good, but I don't know if this works with e.g. firefox bookmarks or ssh connections. I have no idea how bookmark or ssh connection titles cloud be matched.

**Comment by friday** [comment]
I guess a completely different solution is needed to cover that. Extensions could be given an option to add their own items to the regular result database, but I think that's very different from how they work now and not something I would try to cover with the current extension API.

**Comment by jzyinq** [comment]
I'm new to Ulauncher and a big fan of [rofi](https://github.com/davatorium/rofi). It's a dealbreaker for me to open a new app or switch to the existing app window without adding any extra keywords.

I'm really digging the versatility of Ulauncher with its extensions but without the ability to define which extensions could be launched without keywords it just takes a bit more mind processing to get what you need.

**Comment by friday** [comment]
As of #1097 getting merged into v6 (just now) you can define "launch triggers" in the manifest and handle them in the new extension class method `on_launch`.

This will let you add entries to the normal search mode that will trigger directly when you press enter (not asking for any user arguments). They must be universally the same for all users of the extension so it should work for the "play" and "pause" cases, but not "Search my bookmarks/files".

Documentation is still lacking, but it's implemented in https://github.com/friday/ulauncher-gnome-settings (v3 branch).

For cases like handling your Firefox bookmarks or files (where you also want to add them to the normal search mode, but the items are not static) we would need to run multiple APIs to cover this. The extension need a way to generate and update a list of results to be added to Ulauncher's search database. We would need scheduling APIs to handle this (see #1063). So it's actually rather complex.

For even more complex situations like the calculator, you will need "input triggers". Those are functionally equivalent (and backwards compatible with) preferences with type "keyword". They will continue to require keywords. I have a plan for making these opt-in "keywordless" via a user preference though. This preference would only show if the extension author provides a regex or similar so we can determine that the query is a match for that extension. And we would need to either make these non-exclusive (so "0" will show both the calculator result and the app/game "0AD"), or a way to prioritize the which extension will take priority if there is overlap (for example graphically drag and drop extensions in the preferences to change the priority). And this will affect performance very negatively if users enable it for slow apps.

**Comment by bertvandepoel** [comment]
I'd still love to see this happen. For example I'd find it very useful if Ulauncher could both search my applications and my running windows. That way if I forgot that I had an application already running somewhere, I would notice because it would be one of the options!

**Comment by mrchantey** [comment]
I've recently moved from windows and love this application! Its made it so much easier to adjust to the linux ecosystem. 
I do miss being able to *just start typing* the name of the git repository i want to open without considering a prefix, as short as it may be. 
For reference the launcher I've moved from is Powertoys, and I found Ulauncher because it came with omakub :)
